Added `mentions:` filter to find messages mentioning specific…

kumbhar-sarthak opened the pull request:
<!-- Describe your pull request here.-->

Fixes: [#34751](https://github.com/zulip/zulip/issues/34751#event-18201790905)

This pull request adds support for a new `mentions:` search filter, allowing users to search for messages that explicitly mention a specific user using the `mentions:user` syntax. Previously, users could only find messages where they themselves were mentioned via `is:mentioned`. To search for messages mentioning other users, one had to manually type the `@**username**`. This update simplifies that process by introducing a dedicated, user-friendly filter. The filter behaves similarly to dm-including: in the UI, including typeahead suggestions. It supports only one user per pill search. The filter excludes silent mentions (`@_**username**`), group mentions (`@**all**`, `@**everyone**`), and wildcard mentions.

### ✨ Key Features

- Adds support for the `mentions:` search operator.
- Works similarly to `dm-including:`:
  - Shows typeahead suggestions.
  - Supports one user per pill.
- Excludes:
  - Silent mentions (`@_**username**`)
  - Group mentions (`@**all**`, `@**everyone**`)
  - Wildcard mentions
- Added custom error support if specific user not mentioned.
- Included new help under the Help menu.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

## Screenshots and screen captures:

<img src="https://github.com/user-attachments/assets/145127d6-420e-4cf3-b4f7-0bd453d99c55" width="400"/>
<img src="https://github.com/user-attachments/assets/d572e687-51ae-46f3-94ba-d3cb5ef71e98" width="400"/>
<img src="https://github.com/user-attachments/assets/941f0460-974e-40fa-8e3d-9380853b0315" width="400"/>
<img src="https://github.com/user-attachments/assets/7bfe728f-ae22-4145-ad97-2b75da238ef0" width="400"/>
<img src="https://github.com/user-attachments/assets/9f044544-9201-441f-af2b-dde3d2a42712" width="400"/>
<img src="https://github.com/user-attachments/assets/4e200cbf-f6c4-4bb7-ba73-64893edb562b" width="400"/>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on zerver/lib/narrow.py:
```diff
@@ -763,6 +766,25 @@ def by_dm_including(
+        except JsonableError:
+            raise BadNarrowOperatorError("unknown user " + str(operand))
+
+        cond = column("rendered_content").ilike(f'%data-user-id="{user.id}"%') & ~column(
```
No, you can't filter on HTML, it's far too slow. Use the `UserMessage` table, like the existing `has:mention` filter.

kumbhar-sarthak replied:
is filter.ts require to update ??

timabbott replied:
I think I've given you the advice you need: A parallel feature you can review the implementation of.

opmkumar commented:
Do we want to search for the group mentions using this operator too, @alya? 
In my opinion, a user should be able to search for group-mentions, too, if we can search for user mentions.
Currently, this PR doesn't support group-mentions searching.

alya commented:
Group mentions will be a follow-up to a follow-up, as noted in #34969. Thanks for checking. :)

timabbott commented:
Tagging as a completion candidate. @kumbhar-sarthak you're welcome to pick this up again, but I'm assuming you don't have time.

kumbhar-sarthak commented:
Thanks @timabbott !!, i would like to work on this and update PR to address outstanding feedbacks and resolve conflict's.

kumbhar-sarthak commented:
hey @timabbott addressed all feedback's, change the filtering logic which i think speeds up filter now. also resolve conflicts and want review on this.

timabbott commented on web/src/filter.ts:
```diff
@@ -223,6 +223,21 @@ function message_matches_search_term(message: Message, operator: string, operand
+            return (
+                message.content.includes(user_id_str) &&
+                !message.content.includes("user-mention silent")
+            );
```
This hacky check is not correct. Don't do hacky imprecise checks. You can parse the HTML instead, a la how `post_process_content` works, to find mention elements and inspect their classes and attributes directly.

timabbott commented on web/src/filter.ts:
```diff
@@ -1108,6 +1133,7 @@ export class Filter {
             "not-topic",
             "dm",
             "dm-including",
+            "mentions",
```
This is wrong. This operator can contain partial conversations. Don't just randomly add items to lists of similar fields, you have to read the function names and comments documenting what these fields mean or you will write nonsense.

kumbhar-sarthak replied:
apologies for that, removed this, i will concern about this and works on it.

timabbott commented on zerver/lib/narrow.py:
```diff
@@ -676,6 +679,25 @@ def by_dm_including(
+            if isinstance(operand, int):
+                user = get_user_by_id_in_realm_including_cross_realm(int(operand), self.realm)
+            else:
+                user = access_user_by_email(self.user_profile, operand, for_admin=False)
```
As discussed above, let's just do user IDs in the server. The old email format is only for backwards-compatibility, which is not required here as this is a new feature.

timabbott commented on zerver/lib/narrow.py:
```diff
@@ -676,6 +679,25 @@ def by_dm_including(
+            raise BadNarrowOperatorError("unknown user " + str(operand))
+
+        cond = (column("user_profile_id", Integer) == user.id) & (
+            column("flags", Integer).op("&")(literal(UserMessage.flags.mentioned.mask)) != 0
```
We should add a comment here being very clear that this is intended to only cover direct personal mentions, so we don't check the other mention flags.

kumbhar-sarthak commented:
Hey @amanagr would up to do review on this?

amanagr commented:
@kumbhar-sarthak thanks for working on this. We should wait for  #36588 to be merged since this PR will change a lot once that is merged.

kumbhar-sarthak commented on web/src/postprocess_content.ts:
```diff
@@ -390,3 +390,38 @@ function process_emoji_only_message(content: DocumentFragment): void {
+
+// Process user mentions in message content was expensive call,
+// we can process only relative part.
+export function process_user_mention(message_content: string, user_id: number): boolean {
```
@amanagr can you review this?

amanagr commented on web/src/filter.ts:
```diff
@@ -196,6 +197,12 @@ function message_matches_search_term(message: Message, term: NarrowTerm): boolea
+
+        case "mentions": {
+            const user = people.get_by_user_id(term.operand);
+
```
Unnecessary extra space.

kumbhar-sarthak commented on web/src/postprocess_content.ts:
```diff
@@ -390,3 +390,38 @@ function process_emoji_only_message(content: DocumentFragment): void {
+
+        if (tag_start !== -1) {
+            const sub_string = message_content.slice(tag_start, index);
+            const class_pos = sub_string.lastIndexOf('class="');
```
i kind of trying to optimize it with DOMParser what i actually observe the template (Document Fragment) element i used but, still we need to add the html content for every message where actual most of time was spend and nothing we can do with it. that gives roughly around 10,000ms to 15,000ms at 4xcpu cooldown.

<img width="1920" height="965" alt="image" src="https://github.com/user-attachments/assets/2062a3b2-946b-4c77-823c-ba9f465ca04b" />
and we need to just create cache for this in message_store for frequent query call.

kumbhar-sarthak replied:
yeah this approach already i state above I'll open discussion.
